File: go/code_smells/bloaters/long_parameter_list.go
package bloaters

import "fmt"

// CreateUser demonstrates long parameter list code smell
func CreateUser(
	username string,
	email string,
	firstName string,
	lastName string,
	age int,
	country string,
	city string,
	address string,
	postalCode string,
	phoneNumber string,
	isActive bool,
	role string,
) error {
	// Validation
	if username == "" || email == "" {
		return fmt.Errorf("username and email are required")
	}

	// Business logic
	fmt.Printf("Creating user: %s (%s)\n", username, email)
	fmt.Printf("Name: %s %s\n", firstName, lastName)
	fmt.Printf("Location: %s, %s, %s, %s\n", address, city, postalCode, country)
	fmt.Printf("Contact: %s\n", phoneNumber)
	fmt.Printf("Age: %d, Role: %s, Active: %t\n", age, role, isActive)

	return nil
}

// Exercise: Refactor this code to reduce the parameter list
// 1. Group related parameters into structs (e.g., Address, PersonalInfo)
// 2. Consider using the Builder pattern
// 3. Think about which parameters are truly required vs optional

File: java/src/main/java/com/refactoring/codesmells/bloaters/LongParameterList.java
package com.refactoring.codesmells.bloaters;

public class LongParameterList {

    public static void createUser(
            String username,
            String email,
            String firstName,
            String lastName,
            int age,
            String country,
            String city,
            String address,
            String postalCode,
            String phoneNumber,
            boolean isActive,
            String role
    ) {
        // Validation
        if (username == null || username.isEmpty() || email == null || email.isEmpty()) {
            throw new IllegalArgumentException("username and email are required");
        }

        // Business logic
        System.out.printf("Creating user: %s (%s)%n", username, email);
        System.out.printf("Name: %s %s%n", firstName, lastName);
        System.out.printf("Location: %s, %s, %s, %s%n", address, city, postalCode, country);
        System.out.printf("Contact: %s%n", phoneNumber);
        System.out.printf("Age: %d, Role: %s, Active: %b%n", age, role, isActive);
    }
}
